app/src/verifier.h

```diff
@@ -173,36 +178,172 @@ namespace scitt::verifier
+      {
+        throw VerificationError("Protected header KID is missing");
+      }
+      auto cose_key_thumb = phdr.tss_map.cose_key->to_sha256_thumb();
+      std::vector<uint8_t> kid_bytes(phdr.kid->begin(), phdr.kid->end());
+      if (!std::equal(
+            cose_key_thumb.begin(),
+            cose_key_thumb.end(),
+            kid_bytes.begin(),
+            kid_bytes.end()))
+      {
+        throw VerificationError("COSE key thumbprint does not match KID");
+      }
+
+      PublicKey key = phdr.tss_map.cose_key->to_public_key();
+
+      // First verify the signature to then trust the structure integrity
+      // and the attestation contained in the protected header.
+      std::span<uint8_t> payload;
+      try
+      {
+        // ignore unknown critical headers due to the presence of a custom TSS
+        // header
+        payload = cose::verify(data, key, true);
+      }
+      catch (const cose::COSESignatureValidationError& e)
+      {
+        throw VerificationError(fmt::format(
+          "Failed to validate cose signature using the COSE key: {}",
+          e.what()));
+      }
+
+      // Verify the attestation report contained in the protected header
+      // against the AMD certificate chain contained in “snp_endorsements”.
+      // see
+      // https://github.com/microsoft/CCF/blob/afc7ef5eca00d413474de47f91a1827f16618de6/src/js/extensions/snp_attestation.cpp#L35
+      ccf::QuoteInfo quote_info = {};
+      quote_info.format = ccf::QuoteFormat::amd_sev_snp_v1;
+      quote_info.quote = phdr.tss_map.attestation.value();
+      quote_info.endorsements = phdr.tss_map.snp_endorsements.value();
+
+      ccf::pal::PlatformAttestationMeasurement measurement = {};
+      ccf::pal::PlatformAttestationReportData report_data = {};
+      std::optional<ccf::pal::UVMEndorsements> parsed_uvm_endorsements;
+
+      try
+      {
+        ccf::pal::verify_snp_attestation_report(
+          quote_info, measurement, report_data);
+      }
+      catch (const std::exception& e)
+      {
+        throw VerificationError(fmt::format(
+          "Failed to validate SNP attestation report: {}", e.what()));
+      }
+
+      // Now check that the attestation report data matches the cose key
+      // This allows us to verify that the enclave knew about the key
+      if (report_data.data.size() < 32)
+      {
+        throw VerificationError(fmt::format(
+          "Report data length is less than 32 bytes: {}",
+          report_data.data.size()));
+      }
+
+      std::vector<uint8_t> cose_key_hash;
+      try
+      {
+        cose_key_hash = phdr.tss_map.cose_key->to_sha256_thumb();
+      }
+      catch (const std::exception& e)
+      {
+        throw VerificationError(
+          fmt::format("Failed to compute COSE key hash: {}", e.what()));
+      }
+
+      // select first 32 bytes of the report data
+      std::span<uint8_t> report_data_span(
+        report_data.data.data(), std::min(report_data.data.size(), size_t(32)));
```

Comment:
It would be a bit stricter and therefore preferable to make sure the rest is zeros.